### PR TITLE
Ensure that from_array creates a copy

### PR DESCRIPTION
**Pull request by phofl**
- [ ] Closes #xxxx
- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

Yikes, I ran into this when working on linear fusion. Testing that gave me a bunch of weird collision errors which are explainable by the missing copy here

### CODE REVIEW

**Comment by davidhassell** [comment]
Hello - is there an issue associated with this? I hope I haven't missed a discussion, but I'm not sure that this is a useful change. Please allow me to try and explain ...

The proposed change to `from_array` looks like it will _always_ copy an array_like input. But what if we don't want to do such a copy, for memory usage and speed concerns, or because we want to leverage in-place changes at some stage down the line? Is it not better leave any copying to the user (e.g. `da.from_array(a.copy())`), or we could provide a `copy=<bool>` keyword argument (defaults to False to preserve backwards compatability) to `from_array`, which will try to call the input's `copy` method, doing nothing if an `AttributeError` is raised? That said, I realise that it's cleaner, more general, and less typing to just pass in a copy and not have the keyword!

If the input is, e.g., a `list`, then it is already copied when it gets converted to a numpy array (which is fine), so we don't want to copy it again, in any event, I think. Some information in the docstring about copying might be all that's required? (I guess something along the lines of "the input is not copied, unless it is a list or tuple, which gets copied into a numpy array").

Thanks,
David

**Comment by phofl** [comment]
This is a pretty bad bug (but no issue associated)

> because we want to leverage in-place changes at some stage down the line?

This is a very strong -1 from me, anything that changes an array inplace that was passed into a Dask Collection is a very very strong anti pattern and **should** never be relied upon, especially since we say our collections are immutable.

>  for memory usage and speed concerns

A copy is pretty cheap and everything that you do afterwards will have a lot more overhead than just copying the array once. If one copy really screws up the performance of your computation, then Dask is probably not the best tool anyway. Additionally, you really shouldn't pass large arrays into from_array anyway, you should load the data from storage.

**Comment by davidhassell** [comment]
Hi. Thanks for responding.

If I have a 10Gb numpy array and I make a copy, on my 16GB laptop, then everything is killed, that can't be right! Isn't Dask exactly the right tool to use my 10GB data? A copy on that sized array on my laptop would take ~1 second. I have thousands of these, so we're really adding up the time.

I appreciate that the in-place change I described is unusual, and could be not recommended, but I can think of workflows that leverage: Instantiate two Dask arrays from the same numpy array, build up different graphs on each, have the option of changing the source in both prior to computation (i.e. before anything has been passed to the scheduler). I don't want to do that at the moment, but I don't think we should disallow it.

It's rightly up to Dask to choose its own copying strategy (fully endorsed by the Python Array API), but whatever its default is, I believe that the user should have options. 

I think that the proposed solution not only takes our options away, but also produces inconsistent results between different input types: Input types without a copy method, input types whose copy method is "shallow", and in input types whose copy method is "deep" would all give different outcomes - i.e. there's no guarantee that a copy is always going into the Dask graph. Better for the user to pass in a copy if they want to (which they can today), guided by better documentation (which could exist tomorrow)

Cheers,
David

**Comment by davidhassell** [comment]
(P.S. Would it be OK to open an issue for this?)

**Comment by phofl** [comment]
> I appreciate that the in-place change I described is unusual, and could be not recommended, but I can think of workflows that leverage: Instantiate two Dask arrays from the same numpy array, build up different graphs on each, have the option of changing the source in both prior to computation (i.e. before anything has been passed to the scheduler). I don't want to do that at the moment, but I don't think we should disallow it.

That would screw with a bunch of different things since the content of your array goes into the graph generation and influences the keys. Again: You really should **not** do that under any circumstances

>If I have a 10Gb numpy array and I make a copy, on my 16GB laptop, then everything is killed, that can't be right! Isn't Dask exactly the right tool to use my 10GB data? A copy on that sized array on my laptop would take ~1 second. I have thousands of these, so we're really adding up the time.

Where are you getting these arrays from? You should use Dask directly to create them instead of putting them into the graph.

> It's rightly up to Dask to choose its own copying strategy (fully endorsed by the Python Array API), but whatever its default is, I believe that the user should have options.

I disagree, mutating the thing inplace has so many potential pitfalls that barely any user will be aware off.

**Comment by davidhassell** [comment]
Hi,

> Where are you getting these arrays from?You should use Dask directly to create them instead of putting them into the graph.

I don't follow this, I'm afraid. Dask is not the only softwre library out there, and my 10GB array could have been made by another piece of software that outputs (say) numpy arrays, I still want to use Dask to process that output (to perform a reduction, for instance).

 > I disagree, mutating the thing inplace has so many potential pitfalls that barely any user will be aware off.

So let the user be aware of these! Numpy has benefits and dangers from it's view/copy approach - they're well advertised and we're all good with that. I really don't see a problem with making it clear ... perhaps the issue that prompted this wouldn't have arisen if there was a description of the input array in the `from_array` docstring. Currently [there is none](https://github.com/dask/dask/blob/2024.11.2/dask/array/core.py#L3474), but how about:

```
        x : array_like
            The array to be converted to a dask array. `list` and `tuple` inputs are copied but 
            for all other input types, any in-place changes made subsequently to *x* will  
            also affect the returned dask array.
```

Could we please open an issue on this, as per the [development guidlines](https://docs.dask.org/en/latest/develop.html#issues)?

Many thanks,
David

**Comment by fjetter** [comment]
> Hi,
> 
> > Where are you getting these arrays from?You should use Dask directly to create them instead of putting them into the graph.
> 
> I don't follow this, I'm afraid. Dask is not the only softwre library out there, and my 10GB array could have been made by another piece of software that outputs (say) numpy arrays, I still want to use Dask to process that output (to perform a reduction, for instance).
> 

This is possible of course and this is also what this API is for. However, you will achieve better performance if you let dask do the generation/data loading. The data copy that is being introduced here is one example why that is better. There are other reasons that are more complex to explain (TLDR if we use a specialized API we can make certain assumption we can't make for the generic `from_array`)

> > I disagree, mutating the thing inplace has so many potential pitfalls that barely any user will be aware off.
> 
> So let the user be aware of these! Numpy has benefits and dangers from it's view/copy approach - they're well advertised and we're all good with that. I really don't see a problem with making it clear ... perhaps the issue that prompted this wouldn't have arisen if there was a description of the input array in the `from_array` docstring. Currently [there is none](https://github.com/dask/dask/blob/2024.11.2/dask/array/core.py#L3474), but how about:

I think there is a conceptual misunderstanding. In the numpy case, these "dangers" or "pitfalls" are mostly there because the API is sometimes not clear and it is easy to make a mistake. Experienced users can still navigate and leverage this.

Dask operates differently. We don't have the concept/distinction of views vs copies. There is no way for users to guarantee that inplace mutations are safe. This mostly comes from us executing things in parallel. The fact that intermediate results are not mutated is a core assumption the dask architecture builds on.

**Comment by davidhassell** [comment]
Hello,

> Dask operates differently. We don't have the concept/distinction of views vs copies. There is no way for users to guarantee that inplace mutations are safe. This mostly comes from us executing things in parallel. The fact that intermediate results are not mutated is a core assumption the dask architecture builds on.

OK, I understand that, but it is not enforced anywhere else in Dask, so why should it be in `from_array`? (notwithstanding the out-of-memory and speed issues I have already highlighted). Consider the following, which quite happily embeds a non-copy of a numpy array in a dask graph:

```python
In [25]: import dask.array as da

In [26]: import numpy as np

In [27]: dx = da.arange(3)

In [28]: x = np.array([4, 5, 6])

In [29]: dy = dx + x

In [30]: dy.compute()
Out[30]: array([4, 6, 8])

In [31]: x[0] = 999

In [32]: dy.compute()
array([999,   6,   8])
```
Dask can't always account for what we choose to put inside a graph. Rather it should it make it clear to users that what they put in (by whatever means) is not a copy, and they should put copies in if they need to.

By automatically copying, you're going to prevent Dask being used tomorrow in cases for which it works perfectly well today, and there won't be a nice error message, rather the session will just get killed when it runs out of memory.

Thanks,
David

**Comment by davidhassell** [comment]
P.P.S. surely it's time for an issue to be opened on this :)

**Comment by phofl** [comment]
> Consider the following, which quite happily embeds a non-copy of a numpy array in a dask graph:

This is incorrect, we are calling from_array under the hood

**Comment by davidhassell** [comment]
> This is incorrect, we are calling from_array under the hood

Interesting - and I get point. _If_ you were to make `from_array` copy numpy arrays (and variously some other things), you can propagate this copying through all operations. Which makes me even more worried :) If `x` is automatically copied, how could I do `dy = dx + x` for a large enough `x`?

> However, you will achieve better performance if you let dask do the generation/data loading

Only if dask is capable of the generation/data loading - it can't do everything :)

If no one else wants to start an issue on this, I shall ...

**Review by fjetter — APPROVED** [review]
The above conversation is mildly off topic. Supporting inplace mutation of intermediate data is certainly off the table. There is nothing we can do to actually prohibit users from this (other than defensively copying which is what we sometimes do anyhow).

This issue is about whether or not we should be copying an array that is passed to us via `from_array` (or implicitly via something like the above example). The argument about required memory on client side is certainly strong but not necessarily sufficient to block this change. 

FWIW We're doing the same thing for the DataFrame API.

> Consider the following, which quite happily embeds a non-copy of a numpy array in a dask graph:

As @phofl already outlined, this example is working because `from_array` is **not** creating a copy. With the changes proposed by this PR, your example would no longer work as you are outlining since `+ x` also goes through `from_array`. In many ways this example is a strong case **for** this change and not against.

> you're going to prevent Dask being used tomorrow in cases for which it works perfectly well today, and there won't be a nice error message, rather the session will just get killed when it runs out of memory.

I think this is happening on a daily basis and it's also not unusual. Just because something is possible doesn't mean that it is intended, nor does it mean that the behavior is feasible to maintain. I believe the example you are showing is not motivating why this feature is useful beyond saving memory.

I currently consider this behavior a bug and want to unblock other work that is depending on this.

Dask has made a lot of progress towards being easier to use in the past months and we often had to make tough decisions, even some that sacrificed performance. I think doing a copy is a safe default behavior.

@davidhassell it seems like you care about this deeply. If you do, I suggest opening an issue to discuss if there are ways to support a non-copy version as optional. There are probably a couple of nontrivial implications to think through.

**Comment by davidhassell** [comment]
Hi @fjetter and @phofl,

> @davidhassell it seems like you care about this deeply. If you do, I suggest opening an issue to discuss if there are ways to support a non-copy version as optional. There are probably a couple of nontrivial implications to think through.

Sorry for the delay in response ... I appreciate you both taking the time to discuss this with me :) I think I will open the issue you suggest, after consulting with some colleagues where I work, and look forward to continuing the interesting debate.

David

**Comment by CSSFrancis** [comment]
@phofl Can you maybe comment on the effects this `copy` might have on memory mapped numpy arrays?  This appears to load memory mapped object completely into RAM which is not ideal in many cases.  I can do some further testing of this but some guidance beforehand might be nice.

**Comment by keflavich** [comment]
This change broke spectral-cube entirely, as we rely on the array not being read into memory.  [This comment](https://github.com/dask/dask/pull/11524#issuecomment-2555806530) highlights the problem: our use case is explicitly to never load the whole array into memory.  Is there a way to use dask without this forced read?

**Comment by CSSFrancis** [comment]
@keflavich [this](https://github.com/dask/dask/issues/11850) thread has some examples that work.

I was going to see if I could add some code to the from_array function to make sure this doesn't keep happening :)

**Pull request merged**
